In [1]:
import pandas as pd
import warnings 
warnings.filterwarnings('ignore')

In [2]:
df=pd.read_csv(r"C:\Users\comp\Downloads\framingham.csv")
df.head()

male  age  education  currentSmoker  cigsPerDay  BPMeds  prevalentStroke  \
0     1   39        4.0              0         0.0     0.0                0   
1     0   46        2.0              0         0.0     0.0                0   
2     1   48        1.0              1        20.0     0.0                0   
3     0   61        3.0              1        30.0     0.0                0   
4     0   46        3.0              1        23.0     0.0                0   

   prevalentHyp  diabetes  totChol  sysBP  diaBP    BMI  heartRate  glucose  \
0             0         0    195.0  106.0   70.0  26.97       80.0     77.0   
1             0         0    250.0  121.0   81.0  28.73       95.0     76.0   
2             0         0    245.0  127.5   80.0  25.34       75.0     70.0   
3             1         0    225.0  150.0   95.0  28.58       65.0    103.0   
4             0         0    285.0  130.0   84.0  23.10       85.0     85.0   

   TenYearCHD  
0           0  
1           0  
2           0  
3           1  
4           0

In [3]:
#Dropping Education column
df_new=df.drop('education',axis=1)

In [4]:
df_new.head()

male  age  currentSmoker  cigsPerDay  BPMeds  prevalentStroke  \
0     1   39              0         0.0     0.0                0   
1     0   46              0         0.0     0.0                0   
2     1   48              1        20.0     0.0                0   
3     0   61              1        30.0     0.0                0   
4     0   46              1        23.0     0.0                0   

   prevalentHyp  diabetes  totChol  sysBP  diaBP    BMI  heartRate  glucose  \
0             0         0    195.0  106.0   70.0  26.97       80.0     77.0   
1             0         0    250.0  121.0   81.0  28.73       95.0     76.0   
2             0         0    245.0  127.5   80.0  25.34       75.0     70.0   
3             1         0    225.0  150.0   95.0  28.58       65.0    103.0   
4             0         0    285.0  130.0   84.0  23.10       85.0     85.0   

   TenYearCHD  
0           0  
1           0  
2           0  
3           1  
4           0

## EDA(Exploratory Data Analysis)

In [9]:
df_new.shape

(4240, 15)

In [10]:
df_new.describe()

male          age  currentSmoker   cigsPerDay       BPMeds  \
count  4240.000000  4240.000000    4240.000000  4211.000000  4187.000000   
mean      0.429245    49.580189       0.494104     9.005937     0.029615   
std       0.495027     8.572942       0.500024    11.922462     0.169544   
min       0.000000    32.000000       0.000000     0.000000     0.000000   
25%       0.000000    42.000000       0.000000     0.000000     0.000000   
50%       0.000000    49.000000       0.000000     0.000000     0.000000   
75%       1.000000    56.000000       1.000000    20.000000     0.000000   
max       1.000000    70.000000       1.000000    70.000000     1.000000   

       prevalentStroke  prevalentHyp     diabetes      totChol        sysBP  \
count      4240.000000   4240.000000  4240.000000  4190.000000  4240.000000   
mean          0.005896      0.310613     0.025708   236.699523   132.354599   
std           0.076569      0.462799     0.158280    44.591284    22.033300   
min           0.000000      0.000000     0.000000   107.000000    83.500000   
25%           0.000000      0.000000     0.000000   206.000000   117.000000   
50%           0.000000      0.000000     0.000000   234.000000   128.000000   
75%           0.000000      1.000000     0.000000   263.000000   144.000000   
max           1.000000      1.000000     1.000000   696.000000   295.000000   

             diaBP          BMI    heartRate      glucose   TenYearCHD  
count  4240.000000  4221.000000  4239.000000  3852.000000  4240.000000  
mean     82.897759    25.800801    75.878981    81.963655     0.151887  
std      11.910394     4.079840    12.025348    23.954335     0.358953  
min      48.000000    15.540000    44.000000    40.000000     0.000000  
25%      75.000000    23.070000    68.000000    71.000000     0.000000  
50%      82.000000    25.400000    75.000000    78.000000     0.000000  
75%      90.000000    28.040000    83.000000    87.000000     0.000000  
max     142.500000    56.800000   143.000000   394.000000     1.000000

In [11]:
df_new.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4240 entries, 0 to 4239
Data columns (total 15 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   male             4240 non-null   int64  
 1   age              4240 non-null   int64  
 2   currentSmoker    4240 non-null   int64  
 3   cigsPerDay       4211 non-null   float64
 4   BPMeds           4187 non-null   float64
 5   prevalentStroke  4240 non-null   int64  
 6   prevalentHyp     4240 non-null   int64  
 7   diabetes         4240 non-null   int64  
 8   totChol          4190 non-null   float64
 9   sysBP            4240 non-null   float64
 10  diaBP            4240 non-null   float64
 11  BMI              4221 non-null   float64
 12  heartRate        4239 non-null   float64
 13  glucose          3852 non-null   float64
 14  TenYearCHD       4240 non-null   int64  
dtypes: float64(8), int64(7)
memory usage: 497.0 KB


In [12]:
df_new.isnull().sum()

male                 0
age                  0
currentSmoker        0
cigsPerDay          29
BPMeds              53
prevalentStroke      0
prevalentHyp         0
diabetes             0
totChol             50
sysBP                0
diaBP                0
BMI                 19
heartRate            1
glucose            388
TenYearCHD           0
dtype: int64

## Handling null values

In [18]:
# Check Skewness to handle missing values
print('The skewness of cigsPerDay: ',df_new['cigsPerDay'].skew())
print('The skewness of BPMeds: ',df_new['BPMeds'].skew())
print('The skewness of totChol: ',df_new['totChol'].skew())
print('The skewness of BMI : ',df_new['BMI'].skew())
print('The skewness of heartRate: ',df_new['heartRate'].skew())
print('The skewness of glucose: ',df_new['glucose'].skew())


The skewness of cigsPerDay:  1.2470523561848126
The skewness of BPMeds:  5.551461818309443
The skewness of totChol:  0.8718805634765354
The skewness of BMI :  0.9821832986950597
The skewness of heartRate:  0.6443718160367701
The skewness of glucose:  6.2149483495346765


In [21]:
#If skewness  is more than 1 we will use median to handle the missing value and if it is less than 1 and greater than -1 
# we will use mean.
df_new['cigsPerDay']=df_new['cigsPerDay'].fillna(df_new['cigsPerDay'].median())
df_new['BPMeds']=df_new['BPMeds'].fillna(df_new['BPMeds'].median())
df_new['totChol']=df_new['totChol'].fillna(df_new['totChol'].mean())
df_new['BMI']=df_new['BMI'].fillna(df_new['BMI'].mean())
df_new['heartRate']=df_new['heartRate'].fillna(df_new['heartRate'].mean())
df_new['glucose']=df_new['glucose'].fillna(df_new['glucose'].median())

In [22]:
df_new.isnull().sum()

male               0
age                0
currentSmoker      0
cigsPerDay         0
BPMeds             0
prevalentStroke    0
prevalentHyp       0
diabetes           0
totChol            0
sysBP              0
diaBP              0
BMI                0
heartRate          0
glucose            0
TenYearCHD         0
dtype: int64

In [23]:
df_new.shape

(4240, 15)

## Splitting data into train test split

In [28]:
#identifying target column
y=df_new[['TenYearCHD']]
x=df_new.drop(y,axis=1)

In [34]:
from sklearn.model_selection import train_test_split
x_train,x_test,y_train,y_test=train_test_split(x,y,test_size=0.2,random_state=42)
x_train.shape,x_test.shape,y_train.shape,y_test.shape

((3392, 14), (848, 14), (3392, 1), (848, 1))

In [36]:
y_train.value_counts()

TenYearCHD
0             2871
1              521
dtype: int64

Our training datset have 2871 observation in the category 0(which means that the person will not get heartattack in next 10 years),where as only 521 observation in the category of 1(which means that the person will  get heartattack in next 10 years)

If we build the model on this data we will get very high accuracy to correct classify the data into category 0 but a very low accuracy to correctly classify the data into category 1. 

Thus , we will first balance the data by bringing more data points in category 1

## Balancing the dataset

In [41]:
from imblearn.over_sampling import SMOTE
sm=SMOTE(random_state=42)
#To fit this object on X_train and Y_train  we can do this using fit_resample() or using fit_resample()
x_train_new,y_train_new=sm.fit_resample(x_train,y_train) 

In [42]:
y_train_new

TenYearCHD
0              0
1              0
2              0
3              0
4              0
...          ...
5737           1
5738           1
5739           1
5740           1
5741           1

[5742 rows x 1 columns]

In [43]:
#lets check targeted columns got balanced or not
y_train_new.value_counts()

TenYearCHD
0             2871
1             2871
dtype: int64

## Building Model

In [40]:
from sklearn.linear_model import LogisticRegression 
lr=LogisticRegression()
model=lr.fit(x_train_new,y_train_new)
print("The mode is built successfully")

The mode is built successfully


## Predict on test case

In [44]:
y_test['Prediction']=model.predict(x_test)

In [45]:
y_test

TenYearCHD  Prediction
1350           1           1
1434           0           0
2500           0           0
1128           0           1
4144           1           0
...          ...         ...
1844           0           0
4178           0           0
4193           1           1
2897           0           0
910            0           0

[848 rows x 2 columns]

## Model Evaluation

In [46]:
from sklearn.metrics import accuracy_score, confusion_matrix,classification_report

In [52]:
print(confusion_matrix(y_test['TenYearCHD'],y_test['Prediction']))

[[441 284]
 [ 68  55]]


In [48]:
print('accuracy score:',accuracy_score(y_test['TenYearCHD'],y_test['Prediction']))

accuracy score: 0.5849056603773585


In [50]:
print("Classification:",classification_report(y_test['TenYearCHD'],y_test['Prediction']))

Classification:               precision    recall  f1-score   support

           0       0.87      0.61      0.71       725
           1       0.16      0.45      0.24       123

    accuracy                           0.58       848
   macro avg       0.51      0.53      0.48       848
weighted avg       0.76      0.58      0.65       848

